### DISCUSSION TITLE
Project 3. DIPY algorithms Optimizations

### DISCUSSION THREAD

**Discussion by tanishka321** [dipy/dipy#3114] [#GSoC/GSoD]
Dear Dipy mentors,
( @skoudoro  @pjsjongsung  )

## Introduction
I Tanishka Nama, am a junior undergraduate pursuing Btech+Mtech from the Indian Institute of Technology, Varanasi (IIT-BHU).  I am interested in working on **Project 3. DIPY algorithms Optimizations**

I did my Summer Machine Learning Internship at [INMAS Lab](https://www.drdo.gov.in/drdo/labs-and-establishments/institute-nuclear-medicine-allied-sciences-inmas), Defence Research and Development Organisation, India , worked on application and optimization of ML Algorithm to analyze neuro psyche performance on Brain anatomical data and Data acquisition for the multitasking research which holds the potential to be utilized in the recruitment of Indian Air-force.

## Project Proposal
Through this project I aim to optimize the existing algorithms by leveraging the capabilities of the OpenMP API as it offers support for multithreading, auto-vectorization, and parallelization techniques, particularly beneficial in high-computation scenarios such as probabilistic tractography/denoising.

Below are the objectives of my project:

- Identification and resolution of performance bottlenecks within high-computation frameworks.
- Conducting benchmarks and comprehensive comparisons to monitor and evaluate the effectiveness of performance enhancements.
- Implementation of multithreading and multiprocessing utilizing OpenMP or MPI to facilitate parallel processing, thereby augmenting efficiency.

Furthermore, during the review of the documentation, it has come to my attention that the latest denoising technique available, specifically Patch2self, dates back to 2020. Consequently, I am inclined to propose the implementation of the [SDnDTI model.](https://arxiv.org/abs/2111.07220) and a comparison with the previous models.

Let me know if both of my proposals align with the community's objective and please guide me if I am heading in the wrong direction. I am immensely motivated to give back even the small but significant contribution to the community as I have been using it since my previous internship. My next aim before the commencement will be to familiarize myself with the codebase and actively engage in discussions to understand the current challenges and opportunities for improvement.

**Comment by pjsjongsung** [comment]
Hello Tanishka! Thank you for your interest in DIPY GSoC.

I am glad that you want to work on optimization. DIPY could surely benefit from it. Perhaps @skoudoro can provide more specific insight on the optimization objectives.

I am a little skeptical about SDnDTI model, regarding how fixed the model input shape is (for example do we need at least 18 dwi and 3 b0s like the pipeline in their paper?) and if a matlab deep learning model is easily transferable to tensorflow. This is also only a valid question only if porting a model with trained weights is possible without consulting the authors.

I see you are already creating PRs, so I look forward to your full proposal!

**Comment by tanishka321** [comment]
Hi,
While diving deeper, realized that the fixed input shape requirement and the potential challenges in transferring a MATLAB model to TensorFlow are indeed important considerations. I completely agree that these are important factors to consider when evaluating the feasibility of integrating the SDnDTI technique into DIPY and I will thoroughly investigate these aspects.
I'm inclined to focus on the optimization objectives, leveraging OpenMP to enhance performance.

**Comment by tanishka321** [comment]
Dear @skoudoro and @pjsjongsung 
I am writing to discuss the possibility of incorporating CuPy into the DIPY Algorithms Optimizations project that I am interested in contributing to. CuPy is similar to NumPy but designed to utilize the power of NVIDIA CUDA, which can significantly enhance the performance of GPU-accelerated computations. By harnessing the parallel computing capabilities of CUDA-enabled GPUs, CuPy can accelerate computations for tasks like matrix operations, element-wise operations, etc. 
I would like to ascertain whether this proposition aligns with the objectives of the DIPY community 
This will be followed by **assessing the compatibility and identifying algorithms** within DIPY that would benefit from GPU acceleration. This could include computationally intensive tasks such as probabilistic tractography and nonrigid registration

**Reply by skoudoro** [reply]
Hi @tanishka321,

We have already a project doing that: cudipy - https://github.com/dipy/cudipy and we would like to keep them apart for now. 
The DIPY algorithms Optimizations project should mainly focus on Cython, OpenMP, SIMD and algorithms.
The goal is to answer this questions: How can we rewrite some of our algorithms in a much smarter way to be faster.   GPU is not always the answer and this project is NOT focus on ai tools like tensorflow, pytorch, etc... there are other projects for that like project 5

I hope it helps and answer your questions. 

ps:  It is also good to know that multiple candidates can be on the same project with different topics/goals. this is not a problem.  if your interest is on AI, you should look at project 5. If it is on algorithms and cpu optimization, you are welcome here!
